rocAL/rocAL/source/readers/image/external_source_reader.cpp

```diff
@@ -0,0 +1,230 @@
+/*
+Copyright (c) 2019 - 2023 Advanced Micro Devices, Inc. All rights reserved.
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+#include <cassert>
+#include <algorithm>
+#include "external_source_reader.h"
+#include <boost/filesystem.hpp>
+
+namespace filesys = boost::filesystem;
+
+ExternalSourceReader::ExternalSourceReader() {
+    _curr_file_idx = 0;
+    _current_file_size = 0;
+    _current_fPtr = nullptr;
+    _loop = false;
+    _file_id = 0;
+    _shuffle = false;
+    _file_count_all_shards = 0;
+    _loop = false;    // loop not supported for external source
+    _end_of_sequence = false;
+}
+
+// return batch_size() for count_items unless end_of_sequence has been signalled.
+unsigned ExternalSourceReader::count_items() {
+    if (_file_mode == FileMode::FILENAME) {
+        if (_end_of_sequence && _file_names_que.empty()) {
```

Comment:
Can we change _file_names_que and _images_data_q similar? 
Please use queue itself.